packages/core/mesh/messaging/src/signal-client/signal-client.ts

```diff
@@ -179,66 +220,86 @@ export class SignalClient implements SignalMethods {
-    messageStream.subscribe(async (message: SignalMessage) => {
-      this._performance.receivedMessages++;
-      await this._onMessage({
-        author: PublicKey.from(message.author),
-        recipient: PublicKey.from(message.recipient),
-        payload: message.payload
-      });
-    });
-
-    // Saving message stream.
-    if (!this._messageStreams.has(peerId)) {
-      this._messageStreams.set(peerId, messageStream);
-    }
-
-    this._ctx.onDispose(() => {
-      messageStream.close();
-      this._messageStreams.delete(peerId);
-    });
+    log('subscribing to messages', { peerId });
+    this._subscribedMessages.add({ peerId });
+    this._reconcileTask!.schedule();
 
     return {
       unsubscribe: async () => {
-        messageStream.close();
+        this._messageStreams.get(peerId)!.close();
         this._messageStreams.delete(peerId);
+        this._subscribedMessages.delete({ peerId });
       }
     };
   }
 
+  private _scheduleReconcileAfterError() {
+    scheduleTask(
+      this._ctx!,
+      () => {
+        this._reconcileTask!.schedule();
+      },
+      ERROR_RECONCILE_DELAY
+    );
+  }
+
   private _setState(newState: SignalState) {
     this._state = newState;
     this._lastStateChange = new Date();
     log('signal state changed', { status: this.getStatus() });
     this.statusChanged.emit(this.getStatus());
   }
 
-  private _initContext() {
-    this._ctx = new Context({
-      onError: (err) => log.catch(err)
+  private _initConnectionCtx() {
+    void this._connectionCtx?.dispose();
+    this._connectionCtx = this._ctx!.derive();
+    this._connectionCtx.onDispose(() => {
+      log('connection context disposed');
+      Array.from(this._swarmStreams.values()).forEach((stream) => stream.close());
+      Array.from(this._messageStreams.values()).forEach((stream) => stream.close());
+      this._swarmStreams.clear();
+      this._messageStreams.clear();
     });
   }
 
   private _createClient() {
+    log('creating client', { host: this._host, state: this._state });
+
     this._connectionStarted = new Date();
     try {
-      this._client = new SignalRPCClient(this._host);
+      this._client = new SignalRPCClient({
+        url: this._host,
+        callbacks: {
+          onConnected: () => {
+            log('socket connected');
+            this._lastError = undefined;
+            this._reconnectAfter = DEFAULT_RECONNECT_TIMEOUT;
+            this._setState(SignalState.CONNECTED);
+            this._clientReady.wake();
+            this._reconcileTask!.schedule();
+          },
+
+          onDisconnected: () => {
+            log('socket disconnected', { state: this._state });
+            if (this._state !== SignalState.CONNECTED && this._state !== SignalState.CONNECTING) {
+              this._incrementReconnectTimeout();
+            }
+            this._setState(SignalState.DISCONNECTED);
+            this._reconnect();
+          },
+
+          onError: (error) => {
```

Comment:
If onError and onDisconnect get called in-sequence does it create the client twice?